### PR TITLE
megatron: fix MegatronPretrainingBatchSampler

### PR DESCRIPTION
**Pull request by amo33**
This PR fixes ``megatron_batch_sampler.py`` and some other codes as follows:

In ``__iter__``, broadcasting indices and yielding indices were fixed using ``torch.distributed.broadcast``.
In ``__init__``, sizes of dataset were calculated by ``flatflow.sys.getsizeof``. 
In order to fix this issue, there was another code implemented as ``blendable_dataset.py``.

In ``blendable_dataset.py``, ``__sizeof__`` is implemented because samplers dataset was instance of ``BlendableDataset``.

Lastly, ``BlendableDataset`` was added in ``megatron_gpt_sft_model.py`` and ``__init__``. 

Moreover, there is one issue to work on ``megatron_batch_sampler.py`` in ``__init__`` where we call ``self.client`` with ``Init``
~~(Currently working on this and need help to fix this)~~.

Updates(11/23) 
Thanks for our team to discuss problem, changed version of sampler design logic is implemented.
Currently there will be no updates for this PR from now on.

### CODE REVIEW

**Comment by amo33** [comment]
@9rum Other PR will be added within following week. 

- [ ] custom parallel state for model parallel group broadcast
- [ ] Profiler for forward time profile

File: flatflow/nemo/collections/nlp/data/language_modeling/megatron/megatron_batch_samplers.py [hunk 1]
```diff
@@ -111,50 +115,47 @@ def __init__(
...
-                micro_batch_size,
-                order,
-                self.rank,
-                seed,
-                heterogeneous,
-                use_flat_shuffle,
-                hidden_size,
-            )
+        if self.pipeline_parallel_rank == 0 and self.tensor_parallel_rank == 0:
+            if self.global_rank == 0:
```
Comment (9rum) [inline flatflow/nemo/collections/nlp/data/language_modeling/megatron/megatron_batch_samplers.py:136]: I think these lines can be simplified to:
```python
self.client.Init(
    global_batch_size,
    micro_batch_size,
    order,
    self.data_parallel_rank,
    seed,
    heterogeneous,
    use_flat_shuffle,
    hidden_size,
    sizes if self.global_rank == 0 else None,
)
```

File: flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 2]
```diff
@@ -330,6 +330,13 @@ def _build_dataset(self, data_cfg, is_train=True):
         if is_train:
             if self.use_flatflow or packed_sequence:
                 num_train_samples_after_blend = sum(len(dataset) for dataset in datasets)
+            if self.use_flatflow:
+                dataset = flatflow.nemo.collections.nlp.data.language_modeling.megatron.BlendableDataset(
+                    datasets = datasets, 
+                    weights = data_cfg.concat_sampling_probabilities, 
+                    size = num_train_samples_after_blend
```
Comment (9rum) [inline flatflow/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:337]: Please use Ruff format.

File: flatflow/nemo/collections/nlp/data/language_modeling/megatron/megatron_batch_samplers.py [hunk 3]
```diff
@@ -166,33 +167,69 @@ def set_epoch(self, epoch: int) -> None:
...
-
-        # Check the last partial batch and see drop_last is set
-        if 0 < len(batch) and not self.drop_last:
-            indices = [batch[i] for i in range(self.data_parallel_rank, len(batch), self.data_parallel_size)]
-            if self.pad_samples_to_global_batch_size:
-                num_pad = self._global_batch_size // self.data_parallel_size - len(indices)
-                indices = indices + [-1] * num_pad
-            yield indices
+        indices = []
+        end = False
```
Comment (9rum) [inline flatflow/nemo/collections/nlp/data/language_modeling/megatron/megatron_batch_samplers.py:171]: I think ``end`` is not required and the below loop can be simplified to:
```python
while True:
    if self.consumed_samples > self.total_length // self.num_data_parallel_group:
        break
    ...
```

**Review by 9rum — CHANGES_REQUESTED** [review]
@amo33 Thanks for filing this PR.
I have left several comments for minor improvements and formatting issues.
Always appreciate for your hard work.

**Comment by amo33** [comment]
@9rum Thanks for your feedback,
PR is updated.

**Review by 9rum — APPROVED** [review]
LGTM

**Pull request merged**